prow/cmd/deck/main.go

```
@@ -407,6 +420,131 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+			if err != nil {
+				logrus.WithError(err).Error("Regexp failed to compile.")
+				continue
+			}
+			for _, a := range artifactNames {
+				if r.MatchString(a) {
+					matches = append(matches, a)
+				}
+			}
+			for _, vName := range viewerNames {
+				viewerCache[vName] = matches
+			}
+		}
+
+		viewsStart := time.Now()
+		lenses := sg.Views(viewerCache)
+		viewsElapsed := time.Since(viewsStart)
+		logrus.Info("Got views in ", viewsElapsed)
+
+		var viewBuf bytes.Buffer
+		type ViewsTemplate struct {
+			Views       []spyglass.Lens
+			Source      string
+			ViewerCache map[string][]string
+		}
+		vTmpl := ViewsTemplate{
+			Views:       lenses,
+			Source:      src,
+			ViewerCache: viewerCache,
+		}
+		t := template.Must(template.ParseFiles("/static/template/spyglass-template.html"))
+		tErr := t.Execute(&viewBuf, vTmpl)
+		if tErr != nil {
+			logrus.WithError(tErr).Error("Error rendering template.")
+		}
+
+		fmt.Fprint(w, viewBuf.String())
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return view page: ", elapsed)
+	}
+}
+
+// handleArtifactView handles requests to load a specific view for a job
+func handleArtifactView(sg *spyglass.Spyglass) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		setHeadersNoCaching(w)
+		w.Header().Set("Content-Type", "application/json")
+		name := r.URL.Query().Get("name")
+		src := r.URL.Query().Get("src")
+		podName := r.URL.Query().Get("podname")
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "failed to read body", http.StatusBadRequest)
+			return
+		}
+		if name == "" {
+			http.Error(w, "missing name query parameter", http.StatusBadRequest)
+			return
+		}
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		viewReq := &spyglass.ViewRequest{}
+		err = json.Unmarshal(body, viewReq)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to marshal request body")
```

shall we return an error?